pkg/urbit/noun/events.c

```diff
@@ -937,7 +900,96 @@ u3e_save(void)
   _ce_patch_free(pat_u);
   _ce_patch_delete();
 
-  _ce_backup();
+  c3_c pax_c[8193];
+  snprintf(pax_c, sizeof(pax_c), "%s/.urb/bhk", u3P.dir_c);
+  if ( c3n == u3e_copy(pax_c) ) {
+    fprintf(stderr, "loom: failed to copy snapshot to %s\r\n", pax_c);
+  }
+}
+
+//! @n (1) Attempt to `dir_c`.
+//! @n (2) Attempt to create north image file in `dir_c`.
+//! @n (3) Attempt to create south image file in `dir_c`.
+//! @n (4) Copy north and south image files to `dir_c` from `u3P.dir_c`.
+c3_o
+u3e_copy(const c3_c* const dir_c)
+{
+  c3_o suc_o = c3n;
+  static c3_i fla_i = O_RDWR | O_CREAT;
+  static const mode_t mod_u = 0666;
+
+  // (1)
+  if ( 0 != mkdir(dir_c, 0700) && EEXIST != errno ) {
+    fprintf(stderr, "loom: failed to create %s: %s\r\n",
+            dir_c, strerror(errno));
+    goto exit;
+  }
+
+  // (2)
+  u3e_image nop_u = { .nam_c = nor_nam_c, .pgs_w = 0 };
+  c3_c pan_c[8193];
+  snprintf(pan_c, sizeof(pan_c), "%s/%s", dir_c, nop_u.nam_c);
+  if ( -1 == (nop_u.fid_i = open(pan_c, fla_i, mod_u)) ) {
+    fprintf(stderr, "loom: failed to open %s: %s\r\n", pan_c, strerror(errno));
+    goto exit;
+  }
+
+  // (3)
+  u3e_image sop_u = { .nam_c = sou_nam_c, .pgs_w = 0 };
+  c3_c pas_c[8193];
+  snprintf(pas_c, sizeof(pas_c), "%s/%s", dir_c, sop_u.nam_c);
+  if ( -1 == (sop_u.fid_i = open(pas_c, fla_i, mod_u)) ) {
+    fprintf(stderr, "loom: failed to open %s: %s\r\n", pas_c, strerror(errno));
+    goto close_north;
+  }
+
+  // (4)
+  if ( (c3y == _ce_image_copy(&u3P.nor_u, &nop_u)) &&
+       (c3y == _ce_image_copy(&u3P.sou_u, &sop_u)) )
+  {
+    suc_o = c3y;
+  }
+
+close_south:
+  close(sop_u.fid_i);
+  if ( c3n == suc_o ) {
+    unlink(pas_c);
+  }
+close_north:
+  close(nop_u.fid_i);
+  if ( c3n == suc_o ) {
+    unlink(pan_c);
+  }
+exit:
+  return suc_o;
+}
+
+//! @n (1)
+c3_o
+u3e_load(const c3_c* const dir_c)
+{
+  if ( !dir_c ) {
+    return c3n;
+  }
+
+  u3e_image nor_u, sou_u;
+
+  if ( (c3n == _ce_image_open(dir_c, &nor_u)) ||
+       (c3n == _ce_image_open(dir_c, &sou_u)) )
+  {
+    fprintf(stderr, "boot: failed to load snapshot in %s\r\n", dir_c);
+    exit(1);
+  }
+
+  u3e_foul();  // (1)
```

Comment:
If the behavior of this function is to load a snapshot from out-of-band, uncorrelated with the page-tracking and increment updates that drive updates to the live, writeable snapshot, then i think you'll want to call `u3e_foul()` after all the pages have been loaded into memory. (As an optimization, we might later want to parameterize  `_ce_image_blit()`'s page tracking and memory protection behavior, to avoid thrashing that state.)

Also worth noting, this function is necessary but not sufficient to restore an out-of-band snapshot. See `u3m_boot()`, or `_cu_realloc()` (called by `u3u_meld()`) for other, necessary loom reinitialization.

Reply:
> Also worth noting, this function is necessary but not sufficient to restore an out-of-band snapshot. See `u3m_boot()`, or `_cu_realloc()` (called by `u3u_meld()`) for other, necessary loom reinitialization.

In this case, then we should probably refactor `u3e_live()` to load a snapshot from an arbitrary directory so that we can simply call `u3m_boot()` when loading an epoch's snapshot.

Reply:
I don't think that's quite right; they're very different operations. `u3e_live()` restores a snapshot in a mutable context: later calls to `u3e_save()` will update that snapshot incrementally with any intervening, persistent state changes. `u3e_load()` restores an immutable snapshot: subsequent calls to `u3e_save()` won't affect it, but a different, mutable snapshot (optionally?).

I think it's reasonable for that distinction to exist in this API, I'm just noting that it will also need to exist at higher levels as well. (And once those higher levels exist, we may find that this new function needs to change, too.)